openmessaging-api/src/main/java/io/openmessaging/api/consumer/Consumer.java

```diff
@@ -0,0 +1,222 @@
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package io.openmessaging.api.consumer;
+
+import io.openmessaging.api.Message;
+import io.openmessaging.api.MessageId;
+import io.openmessaging.exception.OMSException;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * A consumer interface.
+ */
+public interface Consumer extends AutoCloseable {
+
+    /**
+     * Get a topic for the consumer
+     *
+     * @return topic for the consumer
+     */
+    String getTopic();
+
+    /**
+     * Get a subscription for the consumer
+     *
+     * @return subscription for the consumer
+     */
+    String getSubscription();
+
+    /**
+     * Unsubscribe the consumer
+     * <p>
+     * This call blocks until the consumer is unsubscribed.
+     *
+     * @throws OMSException
+     */
+    void unsubscribe() throws OMSException;
+
+    /**
+     * Asynchronously unsubscribe the consumer
+     *
+     * @return {@link CompletableFuture} for this operation
+     */
+    CompletableFuture<Void> unsubscribeAsync();
+
+    /**
+     * Receives a single message.
+     * <p>
+     * This calls blocks until a message is available.
+     *
+     * @return the received message
+     */
+    Message receive() throws OMSException;
+
+    /**
+     * Receive a single message
+     * <p>
+     * Retrieves a message when it will be available and completes {@link CompletableFuture} with received message.
+     * </p>
+     * <p>
+     * {@code receiveAsync()} should be called subsequently once returned {@code CompletableFuture} gets complete with
+     * received message. Else it creates <i> backlog of receive requests </i> in the application.
+     * </p>
+     *
+     * @return {@link CompletableFuture}<{@link Message}> will be completed when message is available
+     */
+    CompletableFuture<Message> receiveAsync();
+
+    /**
+     * Receive a single message
+     * <p>
+     * Retrieves a message, waiting up to the specified wait time if necessary.
+     *
+     * @param timeout
+     *            0 or less means immediate rather than infinite
+     * @param unit
+     * @return the received {@link Message} or null if no message available before timeout
+     */
+    Message receive(int timeout, TimeUnit unit) throws OMSException;
+
+    /**
+     *
+     * @param listener
+     * @throws OMSException
+     */
+    void receive(MessageListener listener) throws OMSException;
+
+    /**
+     * Acknowledge the consumption of a single message
+     *
+     * @param message
+     *            The {@code Message} to be acknowledged
+     */
+    void ack(Message message) throws OMSException;
```

Comment:
We provide the ack mechanism in [MessageListener.Context](http://openmessaging.cloud/openmessaging-java/io/openmessaging/consumer/MessageListener.Context.html) rather than in `Consumer`, we could consider which one is better. 😄

Reply:
check `ackCummulative` and `ack`. it makes more sense to put `ack` in consumer, when you want to `ack` and `ackCummulative`. If you put `ack` in MessageListener.Context, you can't do ack cumulatively. There should be an order implication on acknowledge.